runtime/lua/vim/fcnotify.lua

```diff
@@ -0,0 +1,225 @@
+-- Create a file watcher, each watcher is identified by the name of the file that it
+-- watches.
+
+local uv = vim.loop
+local Watcher = {}
+Watcher.__index = Watcher
+local w = {}
+local WatcherList = {}
+local check_handle = nil
+
+-- Callback for the check handle, checks if there are pending notifications
+-- for any watcher, and handles them as per the value of the `fcnotify`
+-- option.
+local function check_notifications()
+  for _, watcher in pairs(WatcherList) do
+    if watcher.pending_notifs == true then
+      vim.api.nvim_command('checktime '..watcher.bufnr)
+      watcher.pending_notifs = false
+    end
+  end
+end
+
+local function fs_event_start(bufnr)
+  WatcherList[bufnr].handle = uv.new_fs_event()
+  WatcherList[bufnr].handle:start(WatcherList[bufnr].ffname, {}, vim.schedule_wrap(function(...)
+    WatcherList[bufnr]:on_change(...)
+  end))
+end
+
+local function check_handle_start()
+  check_handle = uv.new_check()
+  check_handle:start(vim.schedule_wrap(check_notifications))
+end
+
+local function set_mechanism(option_type, bufnr)
+  if option_type == 'global' then
+    local option = vim.api.nvim_get_option('filechangenotify')
+    if option:find('watcher') then
+      for _, watcher in pairs(WatcherList) do
+        watcher._start_handle = fs_event_start
+      end
+      Watcher.start_notifications = check_handle_start
+    else
+      for _, watcher in pairs(WatcherList) do
+        watcher._start_handle = function(_) end
+      end
+      Watcher.start_notifications = function(_) end
+    end
+  elseif option_type == 'local' then
+    bufnr = bufnr or vim.api.nvim_get_current_buf()
+    local status, option = pcall(vim.api.nvim_buf_get_option, bufnr, 'filechangenotify')
+    if not status then
+      option = vim.api.nvim_get_option('filechangenotify')
+    end
+    if option:find('watcher') then
+      WatcherList[bufnr]._start_handle = fs_event_start
+      Watcher.start_notifications = check_handle_start
+    else
+      WatcherList[bufnr]._start_handle = function(_) end
+    end
+  end
+  Watcher.stop_notifications()
+  Watcher.start_notifications()
+end
+
+-- Checks if a buffer should have a watcher attached to it.
+local function valid_buf(bufnr)
+  if bufnr < 0 then
+    return false
+  end
+
+  local fname = vim.api.nvim_buf_get_name(bufnr)
+  local buflisted = vim.api.nvim_buf_get_option(bufnr, 'buflisted')
+  local buftype = vim.api.nvim_buf_get_option(bufnr, 'buftype')
+
+  if not buflisted or buftype == 'nofile' or buftype == 'quickfix'
+    or not fname then
+    return false
+  end
+  return true
```

Comment:
You change just return the condition, rather than doing an if statement here.

`return buflisted or ...`